### DISCUSSION TITLE
Spring Modulith and Gradle Modules

### DISCUSSION THREAD

**Discussion by FKorni2** [spring-projects/spring-modulith#151] [Q&A]
I am very interested in the Spring Modulith project as it contains quite some things I would like to apply to our project. Yet, I am unsure if it would work with our project setup.

To my understanding Spring Modulith defines the domain structure only by package structure. The examples I saw are rather simple and only small projects. Now that I have a quite big application that not only consists of one module (in terms of gradle module/project) but of many. I am not sure if that works out or if it even is designed for such a multi module project.

**Comment by kriehzoo** [comment]
I am trying the same thing with a Maven multi-module project, having each application module as Maven module plus one solution module that packages all the modules together (Spring fat jar).

One issue I am facing is that I have to put my application module tests into my solution module. Because if I put them into the module where they would belong, the test cannot be bootstrapped. More specifically, it fails with:

`java.lang.IllegalArgumentException: Modulith root type must not be null!`

**Comment by mariusz-sieraczkiewicz** [comment]
Exactly, this is the most important feature to be provided, for any non-trivial project application modules should be separate gradle modules, otherwise it would not be manageable on larger scale.

**Comment by apetrelli** [comment]
Just a starter, I think that there is not anything OOTB, however I think that the possibility of achieving multi-jar modules in Spring Modulith relies on the customization of the module detection:
https://docs.spring.io/spring-modulith/reference/fundamentals.html#customizing-modules
